### PR TITLE
Detect usage of ICryptoTransform that would be thread-unsafe

### PR DESCRIPTION
**Pull request by raulgarciamsft**
Adding a new rule for detecting usage of static objects that implement ICryptoTransform that would be thread-unsafe, and potentially result in incorrect cryptographic results.

### CODE REVIEW

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+class ICryptoTransform extends Class {
+  ICryptoTransform() {
+    this.getABaseType*().hasQualifiedName("System.Security.Cryptography", "ICryptoTransform")
+  }
+}
+
+predicate usesICryptoTransformType( Type t ) {
+  exists(  ICryptoTransform ict |
+    ict = t
+    or usesICryptoTransformType( t.getAChild*() )
```
Comment (hvitved) [inline csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql]: Since this predicate is recursive anyway, you can omit the `*`.

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  )
+}
+
+predicate hasICryptoTransformMember( Class c) {
+  exists( Field f |
+    f = c.getAMember*()
```
Comment (hvitved) [inline csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql]: I believe the `*` is redundant.

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        )
+      )
+    )
+  )
+  or
+  exists( Field f |
+    f = c.getAMember*()
+    and not f.isStatic()
+    and ( hasICryptoTransformStaticMember( f.getType(), _ )
+      and msg = "Non-static field " + f + " of type " + f.getType() + " internally makes use of an static object that implements 'System.Security.Cryptography.ICryptoTransform'. This causes that usage of this class member is unsafe for concurrent threads."
```
Comment (hvitved) [inline csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql]: Perhaps this case should be limited to situations where the underlying static field is not in source code? Otherwise I would expect the underlying field to have already been flagged by one of the cases above.

Comment (hvitved) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql]: Also, an static -> a static

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp [hunk 4]
```diff
@@ -0,0 +1,34 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>Classes that implement <code>System.Security.Cryptography.ICryptoTransform</code> are not thread safe.</p>
```
[the remark this answers is not in the document]
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp]: Unfortunately the documentation in Microsoft docs is no longer specific about the instance members not being thread-safe.
The problem is because of the way the class is implemented using CAPI/CNG patterns. In an example of a hash class implementing this interface, typically there would be an instance specific hash object (say using BCryptCreateHash function), which can be called multiple times to add data to the hash (i.e. BCryptHashData), and finally calling the function that would finish the hash & get back the data (i.e. BCryptFinishHash). 
This implementation would potentially allow the same hash object to be called with data from multiple threads before calling the finish function, thus leading to potentially incorrect results.
For example, on a test I have multiple threads hashing "abc" on a static hash function, at least in one occasion I got the result for "abcabc".
It is very unlikely somebody outside Microsoft would write a class that implements ICryptoTransform, and even if they do, it is likely that they will follow the same CAPI/CNG pattern.

File: csharp/ql/src/Likely Bugs/ThreadUnSafeICryptoTransformFix.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
+internal class TokenCacheThreadUnsafeICryptoTransformDemo
```
[the remark this answers is not in the document]
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnSafeICryptoTransformFix.cs]: Fixed.

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql (hunk 1 above)
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql]: Fixed

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql (hunk 2 above)
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql]: Confirmed & fixed

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql (hunk 3 above)
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql]: If I remove this one, as you mentioned the underlying field class would have been flagged anyway.
I was looking for corner cases, but perhaps it is cleaner without it.
Giving it a try without this case.

Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql]: Please let me know what do you think about this one, or if I should add something that looks for cases where the underlying field is not in source code (the culprit class should still be flagged either way).

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp (hunk 4 above)
[the remark this answers is not in the document]
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp]: I forgot to ask. Do you want me to add this explanation (after a cleanup) on the .qhelp file?

[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp]: That would indeed be a nice addition to the qhelp, thanks.

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql (hunk 3 above)
Comment (hvitved) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql]: I will let you decide, either remove this case, or restrict it to situations where the underlying static field is not in source code.

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp (hunk 4 above)
[the remark this answers is not in the document]
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp]: Fixed. I expanded the documentation. Please feel free to correct (or let me know if I should correct) any typo or grammatical error.

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql (hunk 3 above)
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql]: I removed the case.
BTW. I am working on a separate QL that will try to find if any Lambda is capturing a local variable that is of type ICryptoTransform (or has a member of this type). The new QL should detect cases that this one cannot such as the stackoverflow example & complement each other. I will probably have it ready next week.

File: csharp/ql/src/Likely Bugs/ThreadUnSafeICryptoTransformBad.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
+internal class TokenCacheThreadUnsafeICryptoTransformDemoFixed
```
Comment (hvitved) [inline csharp/ql/src/Likely Bugs/ThreadUnSafeICryptoTransformBad.cs]: This file should be named `ThreadUnSafeICryptoTransformGood.cs`.

File: csharp/ql/src/Likely Bugs/ThreadUnSafeICryptoTransformGood.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
+internal class TokenCacheThreadUnsafeICryptoTransformDemo
```
Comment (hvitved) [inline csharp/ql/src/Likely Bugs/ThreadUnSafeICryptoTransformGood.cs]: This file should be named `ThreadUnSafeICryptoTransformBad.cs`.

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+<p>The implementation would potentially allow the same hash object to be called with data from multiple threads before calling the finish function, thus leading to potentially incorrect results.</p>
+<p>Because of this pattern, you can expect that, if you have multiple threads hashing <code>"abc"</code> on a static hash object, you may occasionally obtain the results (incorrectly) for hashing <code>"abcabc"</code>, or face other unexpected behavior.</p>
+<p>It is very unlikely somebody outside Microsoft would write a class that implements <code>ICryptoTransform</code>, and even if they do, it is likely that they will follow the same common pattern than the existing classes implementing this interface.</p>
+<p>Any object that implements <code>System.Security.Cryptography.ICryptoTransform</code> should not be used in concurrent threads as the instance members of such object are also not thread safe.</p>
```
Comment (jf205) [inline csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp]: '...instance members of such _an_ object are also not thread safe.'

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+<p>Potential problems may not be evident at first, but can range from explicit errors such as exceptions, to incorrect results when sharing an instance of such object in multiple threads.</p>
+
+</overview>
+<recommendation>
+<p>Verify that the object is not being shared across threads.</p>
+<p>If it is shared accross instances. Consider changing the code to use a non-static object of type <code>System.Security.Cryptography.ICryptoTransform</code> instead.</p>
```
Comment (jf205) [inline csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp]: I suggest joining these two sentences:

<p>If the object is shared across instances, you should consider changing the code to use a non-static object of type <code>System.Security.Cryptography.ICryptoTransform</code> instead.</p>

Also a typo: _across_ not _accross_

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+<p>As an alternative, you could also look into using <code>ThreadStatic</code> attribute, but make sure you read the initialization remarks on the documentation.</p>
+
+</recommendation>
+<example>
+<p>This example demonstrates the dangers of using a static <code>System.Security.Cryptography.ICryptoTransform</code> in such a way that the results may be incorrect.</p>
```
Comment (jf205) [inline csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp]: Maybe rephrase:

<p>This example demonstrates the dangers of using a static <code>System.Security.Cryptography.ICryptoTransform</code> in a way that generates incorrect results.</p>

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp [hunk 11]
```diff
@@ -0,0 +1,39 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>Classes that implement <code>System.Security.Cryptography.ICryptoTransform</code> are not thread safe.</p>
+<p>The root cause of the problem is because of the way these class are implemented using Microsoft CAPI/CNG patterns.</p>
```
[the remark this answers is not in the document]
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp]: Fixed

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp (hunk 8 above)
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp]: Fixed

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp (hunk 9 above)
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp]: Fixed

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp (hunk 10 above)
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp]: Fixed

File: csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+<p>For example, a hash class implementing this interface, typically there would be an instance-specific hash object (i.e. using <code>BCryptCreateHash</code> function), which can be called multiple times to add data to the hash (i.e. <code>BCryptHashData</code>), and finally calling the function that would finish the hash & get back the data (i.e. <code>BCryptFinishHash</code>).</p>
+<p>The implementation would potentially allow the same hash object to be called with data from multiple threads before calling the finish function, thus leading to potentially incorrect results.</p>
+<p>Because of this pattern, you can expect that, if you have multiple threads hashing <code>"abc"</code> on a static hash object, you may occasionally obtain the results (incorrectly) for hashing <code>"abcabc"</code>, or face other unexpected behavior.</p>
+<p>It is very unlikely somebody outside Microsoft would write a class that implements <code>ICryptoTransform</code>, and even if they do, it is likely that they will follow the same common pattern than the existing classes implementing this interface.</p>
```
[the remark this answers is not in the document]
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.qhelp]: fixed

File: csharp/ql/src/Likely Bugs/ThreadUnSafeICryptoTransformBad.cs (hunk 6 above)
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnSafeICryptoTransformBad.cs]: Fixed it.

File: csharp/ql/src/Likely Bugs/ThreadUnSafeICryptoTransformGood.cs (hunk 7 above)
Comment (raulgarciamsft) [reply csharp/ql/src/Likely Bugs/ThreadUnSafeICryptoTransformGood.cs]: Fixed it.

**Comment by jf205** [comment]
The updated documentation LGTM. Thanks.
